hsm-samples/go/hsm-sample.go

```diff
@@ -0,0 +1,228 @@
+		if err != nil {
+			panic(err)
+		}
+		panic(fmt.Errorf("transaction %s failed to commit with status code: %d", commit.TransactionID(), int32(status)))
+	}
+
+	fmt.Printf("Transaction committed successfully\n")
+	fmt.Println("Evaluating \"get\" query with arguments: async")
+
+	evaluateResult, err := contract.EvaluateTransaction("get", "async")
+	if err != nil {
+		panic(fmt.Errorf("failed to evaluate transaction: %w", err))
+	}
+
+	fmt.Printf("Query result = %s\n", string(evaluateResult))
+}
+
+// newGrpcConnection creates a gRPC connection to the Gateway server.
+func newGrpcConnection() *grpc.ClientConn {
+	certificate, err := loadCertificate(tlsCertPath)
+	if err != nil {
+		panic(fmt.Errorf("failed to obtain commit status: %w", err))
+	}
+
+	certPool := x509.NewCertPool()
+	certPool.AddCert(certificate)
+	transportCredentials := credentials.NewClientTLSFromCert(certPool, "peer0.org1.example.com")
+
+	connection, err := grpc.Dial(peerEndpoint, grpc.WithTransportCredentials(transportCredentials))
+	if err != nil {
+		panic(fmt.Errorf("failed to evaluate transaction: %w", err))
+	}
+
+	return connection
+}
+
+// newIdentity creates a client identity for this Gateway connection using an X.509 certificate.
+func newIdentity(certificatePEM []byte) *identity.X509Identity {
+	cert, err := identity.CertificateFromPEM(certificatePEM)
+	if err != nil {
+		panic(err)
+	}
+	id, err := identity.NewX509Identity(mspID, cert)
+	if err != nil {
+		panic(err)
+	}
+
+	return id
+}
+
+// newSign creates a function that generates a digital signature from a message digest using a private key.
+func newHSMSign(h *identity.HSMSignerFactory, certPEM []byte) (identity.Sign, identity.HSMSignClose) {
+	opt := identity.HSMSignerOptions{
+		Label:      "ForFabric",
+		Pin:        "98765432",
+		Identifier: string(certPEM),
+	}
+
+	sign, close, err := h.NewHSMSigner(opt)
+	if err != nil {
+		panic(err)
+	}
+
+	return sign, close
+}
+
+func loadCertificate(filename string) (*x509.Certificate, error) {
+	certificatePEM, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	return identity.CertificateFromPEM(certificatePEM)
+}
+
+func getSKI(certPEM []byte) []byte {
+	block, _ := pem.Decode(certPEM)
+
+	x590cert, _ := x509.ParseCertificate(block.Bytes)
+	pk := x590cert.PublicKey
+
+	return skiForKey(pk.(*ecdsa.PublicKey))
+}
+
+func skiForKey(pk *ecdsa.PublicKey) []byte {
+	ski := sha256.Sum256(elliptic.Marshal(pk.Curve, pk.X, pk.Y))
+	return ski[:]
+}
+
+func findSoftHSMLibrary() string {
+
+	libraryLocations := []string{
+		"/usr/lib/softhsm/libsofthsm2.so",
+		"/usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so",
+		"/usr/local/lib/softhsm/libsofthsm2.so",
+		"/usr/lib/libacsp-pkcs11.so",
+	}
+
+	for _, libraryLocation := range libraryLocations {
+		if _, err := os.Stat(libraryLocation); errors.Is(err, os.ErrNotExist) {
+			// file does not exist
+		} else {
+			return libraryLocation
+		}
+	}
+	return "NoLibraryFound"
```

Comment:
This doesn't look like very best-practice error handling. Probably this function should either return `(string, error)` or (since it's just a sample) just panic if no library exists

Reply:
panic seems the best option here